src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs

```diff
@@ -1229,12 +1249,28 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
                     GetOriginalMethods(result), methodGroup.ResultKind, methodGroup.TypeArguments.ToImmutable(), analyzedArguments, invokedAsExtensionMethod: invokedAsExtensionMethod, isDelegate: ((object)delegateTypeOpt != null));
             }
 
-            // Otherwise, there were no dynamic arguments and overload resolution found a unique best candidate. 
+            // Otherwise, overload resolution found a unique best candidate. 
             // We still have to determine if it passes final validation.
 
             var methodResult = result.ValidResult;
             var returnType = methodResult.Member.ReturnType;
             var method = methodResult.Member;
+            bool forceDynamicResultType = false;
+
+            var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+
+            if (hasDynamicArgument && !methodGroup.IsExtensionMethodGroup && method.MethodKind != MethodKind.LocalFunction &&
+                !method.ReturnsVoid && !method.ReturnsByRef && !returnType.IsDynamic() &&
+                Conversions.ClassifyConversionFromExpressionType(returnType, Compilation.DynamicType, isChecked: false, ref useSiteInfo).IsImplicit &&
+                !HasApplicableMemberWithPossiblyExpandedNonArrayParamsCollection(analyzedArguments.Arguments, ImmutableArray.Create(methodResult)))
+            {
+                var tryDynamicInvocationDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
+                BindDynamicInvocation(node, targetMethodGroupOrDelegateInstance, analyzedArguments, ImmutableArray.Create(method), tryDynamicInvocationDiagnostics, queryClause);
+                forceDynamicResultType = !tryDynamicInvocationDiagnostics.HasAnyResolvedErrors();
```

Comment:
Consider adding a comment to summarize intent: for scenarios that bound dynamically prior to C# 13, we force a dynamic result type despite compile-time resolution of the call for back compat.